src/action-providers.ts

```diff
@@ -21,18 +21,87 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+
+    const onCanPeform = (
+      path: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => {
+      if (isLegacyActionProvider(refactoring.actionProvider)) {
+        return;
+      }
+
+      if (refactoring.actionProvider.updateMessage) {
+        refactoring.actionProvider.updateMessage(path);
+      }
+
+      applicableRefactorings.push(refactoring);
+    };
+
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(refactoring =>
+          this.canPerform(refactoring, path, selection, onCanPeform)
+        );
+      }
+    });
+
+    const applicableLegacyRefactorings = this.refactorings.filter(refactoring =>
+      this.canPerformLegacy(refactoring, ast, selection)
+    );
+
+    return [...applicableRefactorings, ...applicableLegacyRefactorings].map(
+      refactoring => this.buildCodeActionFor(refactoring)
+    );
   }
 
   private canPerform(
+    refactoring: RefactoringWithActionProvider,
+    path: t.NodePath<any>,
+    selection: Selection,
+    onCanPerform: (
+      matchedPath: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => void
+  ) {
+    if (isLegacyActionProvider(refactoring.actionProvider)) {
+      return;
+    }
+
+    const visitor: t.Visitor = refactoring.actionProvider.createVisitor(
+      selection,
+      path => onCanPerform(path, refactoring),
+      refactoring
+    );
+
+    this.visit(visitor, path);
+  }
+
+  private visit(visitor: any, path: t.NodePath<any>) {
```

Comment:
I think `NodePath` will default on `t.Node` if we don't provide a value to the generics:

```suggestion
  private visit(visitor: any, path: t.NodePath) {
```

Reply:
Also: is `visitor` really `any`thing? It seems it should be `t.Visitor`, isn't it?

Reply:
It is not really anything but it can mean a lot of things and it has lots of generic stuff. I have tried to use `t.Visitor` instead but this leads to a lot of unreadable code. You can give it a try. Your typescript magic might be stronger than mine. 😄